lib/recursive-methods.rb

```diff
@@ -1,49 +1,136 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError.new("Must be a positive integer.")
+  elsif n == 0 
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return s
+  end
+  
+  i = (s.length - 1)
+  
+  if i == 0
+    return s
+  else
+    return s[i] + reverse(s[0..-2])
```

Comment:
`s[0..-2]` makes a new array and copies elements from `s` over to the new array.